Index: linux-2.6.32.67/kernel/panic.c
===================================================================
--- linux-2.6.32.67.orig/kernel/panic.c
+++ linux-2.6.32.67/kernel/panic.c
@@ -71,9 +71,7 @@ NORET_TYPE void panic(const char * fmt,
 	vsnprintf(buf, sizeof(buf), fmt, args);
 	va_end(args);
 	printk(KERN_EMERG "Kernel panic - not syncing: %s\n",buf);
-#ifdef CONFIG_DEBUG_BUGVERBOSE
 	dump_stack();
-#endif
 
 	kmsg_dump(KMSG_DUMP_PANIC);
 	/*
