--- a/drivers/pci/setup-irq.c
+++ b/drivers/pci/setup-irq.c
@@ -30,15 +30,15 @@ pdev_fixup_irq(struct pci_dev *dev,
 	   which interrupt pin it will come in on.   We know which slot it
 	   will come in on 'cos that slot is where the bridge is.   Each
 	   time the interrupt line passes through a PCI-PCI bridge we must
 	   apply the swizzle function.  */
 
 	pci_read_config_byte(dev, PCI_INTERRUPT_PIN, &pin);
 	/* Cope with illegal. */
-	if (pin > 4)
+	if (pin == 0 || pin > 4)
 		pin = 1;
 
 	if (pin != 0) {
 		/* Follow the chain of bridges, swizzling as we go.  */
 		slot = (*swizzle)(dev, &pin);
 
 		irq = (*map_irq)(dev, slot, pin);
