--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -535,14 +535,17 @@ static int hostapd_ctrl_iface_nfc_report
 static int hostapd_ctrl_iface_wps_ap_pin(struct hostapd_data *hapd, char *txt,
 					 char *buf, size_t buflen)
 {
 	int timeout = 300;
 	char *pos;
 	const char *pin_txt;
 
+	if (!hapd->wps)
+		return -1;
+
 	pos = os_strchr(txt, ' ');
 	if (pos)
 		*pos++ = '\0';
 
 	if (os_strcmp(txt, "disable") == 0) {
 		hostapd_wps_ap_pin_disable(hapd);
 		return os_snprintf(buf, buflen, "OK\n");
