Index: quilt-0.47/configure.ac
===================================================================
--- quilt-0.47.orig/configure.ac
+++ quilt-0.47/configure.ac
@@ -274,7 +274,11 @@ fi
 AC_MSG_CHECKING([the version of $PATCH])
 if $PATCH --version 2> /dev/null | grep GNU >/dev/null; then
   set -- `$PATCH --version 2> /dev/null`
-  patch_version=$2
+  if test x$1 = xGNU ; then
+    patch_version=$3
+  else
+    patch_version=$2
+  fi
   AC_MSG_RESULT($patch_version)
   saved_IFS=$IFS; IFS='.'
   set -- $patch_version
Index: quilt-0.47/configure
===================================================================
--- quilt-0.47.orig/configure
+++ quilt-0.47/configure
@@ -5896,7 +5896,11 @@ fi
 echo $ECHO_N "checking the version of $PATCH... $ECHO_C" >&6; }
 if $PATCH --version 2> /dev/null | grep GNU >/dev/null; then
   set -- `$PATCH --version 2> /dev/null`
-  patch_version=$2
+  if test x$1 = xGNU ; then
+    patch_version=$3
+  else
+    patch_version=$2
+  fi
   { echo "$as_me:$LINENO: result: $patch_version" >&5
 echo "${ECHO_T}$patch_version" >&6; }
   saved_IFS=$IFS; IFS='.'
